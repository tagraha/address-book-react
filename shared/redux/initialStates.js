export const usersMock = {
  page: 0,
  seed: 'tirta',
  filterKeyword: '',
  data: {
    isLoading: false,
    isError: false,
    info: {},
    results: [],
    fetchedCount: 0,
  },
  filteredUsers: {
    isShown: false,
    results: [],
  },
  config: {
    nationalities: [
      {
        countryCode: 'AU',
        isChecked: true,
      },
      {
        countryCode: 'BR',
        isChecked: false,
      },
      {
        countryCode: 'CA',
        isChecked: false,
      },
      {
        countryCode: 'CH',
        isChecked: false,
      },
      {
        countryCode: 'DE',
        isChecked: false,
      },
      {
        countryCode: 'ES',
        isChecked: false,
      },
      {
        countryCode: 'FI',
        isChecked: false,
      },
      {
        countryCode: 'FR',
        isChecked: false,
      },
      {
        countryCode: 'GB',
        isChecked: false,
      },
      {
        countryCode: 'IE',
        isChecked: false,
      },
      {
        countryCode: 'IR',
        isChecked: false,
      },
      {
        countryCode: 'NO',
        isChecked: false,
      },
      {
        countryCode: 'NL',
        isChecked: false,
      },
      {
        countryCode: 'NZ',
        isChecked: false,
      },
      {
        countryCode: 'TR',
        isChecked: false,
      },
      {
        countryCode: 'US',
        isChecked: false,
      },
    ],
  },
}

export const postMock = {
  asyncPostExample: {
    body: '',
    id: 0,
    title: '',
    userId: 0,
  },
};

export const counterMock = {
  counterValue: 0,
};

export const errorMock = {};
