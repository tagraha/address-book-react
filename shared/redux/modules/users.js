// users.js
import { usersMock } from './../initialStates';

// Actions
const FETCH_USERS = 'users/FETCH_USERS';
const FETCH_USERS_SUCCESS = 'users/FETCH_USERS_SUCCESS';
const FETCH_USERS_FAILED = 'users/FETCH_USERS_FAILED';

/**
 * Helper function to calculate fetched users data
 * @constructor
 * @param {string} current - The number of the current fetched users data.
 * @param {string} addition - The number of new fetched users data count.
 */
function countFetchedData(current, addition) {
  return parseInt(current + addition)
}

// Reducer
// DUCKS: https://github.com/erikras/ducks-modular-redux
export default function reducer(state = usersMock, action = {}) {
  switch (action.type) {
    // do reducer stuff
    case FETCH_USERS: {
      return {
        ...state,
        data: {
          // only change isLoading data object key
          // the rest object key is not affected
          ...state.data,
          isLoading: true,
        }
      };
    }
    case FETCH_USERS_SUCCESS: {
      // get the current users data array on redux state tree
      const prevUsersData = state.data.results;
      // fresh fetched users data from API
      const fetchedData = action.payload.results;
      // combine the users data
      const usersData = prevUsersData.concat(fetchedData);

      return {
        ...state,
        page: action.payload.info.page,
        data: {
          ...state.data,
          isError: false,
          isLoading: false,
          info: action.payload.info,
          results: usersData,
          fetchedCount: countFetchedData(state.data.fetchedCount, action.payload.info.results),
        }
      };
    }
    case FETCH_USERS_FAILED: {
      return {
        ...state,
        data: {
          ...state.data,
          isLoading: false,
          isError: true,
        }
      };
    }
    default: {
      return state;
    }
  }
}

// Action Creators
function fetchUsers() {
  return { type: FETCH_USERS };
}

function fetchUsersSuccess(users) {
  return { type: FETCH_USERS_SUCCESS, payload: users };
}

function fetchUsersFailed(err) {
  return { type: FETCH_USERS_FAILED, payload: err.message };
}

// side effects, only as applicable
// e.g. thunks, epics, etc
export const loadUsers = (results = 50) => (dispatch, getState, { axios }) => {
  dispatch(fetchUsers());
  // set current pagination state tree
  const pagination = parseInt(getState().users.page);
  return (
    axios
      .get(`https://randomuser.me/api/?page=${pagination+1}&results=${results}&seed=tirta`)
      .then((res) => {
        dispatch(fetchUsersSuccess(res.data));
      })
      // eslint-disable-next-line
      .catch((err) => {
        dispatch(fetchUsersFailed(err));
      })
  );
}
