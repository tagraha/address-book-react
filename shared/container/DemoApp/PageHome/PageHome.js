import React from 'react';
import Helmet from 'react-helmet';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { withJob } from 'react-jobs';
import debounce from 'lodash.debounce';

import { loadInitialUsers, loadUsers } from './../../../redux/modules/users';
import UsersCard from '../../../components/UsersCard/UsersCard';

import config from '../../../../config';

class PageHome extends React.Component {
  constructor(props) {
    super(props);
    this.handleOnScroll = this.handleOnScroll.bind(this);
  }
  componentDidMount() {
    window.addEventListener('scroll', this.handleOnScroll);
  }
  componentWillUnmount() {
    window.removeEventListener('scroll', this.handleOnScroll);
  }
  handleOnScroll() {
    var scrollTop = (document.documentElement && document.documentElement.scrollTop) || document.body.scrollTop;
    var scrollHeight = (document.documentElement && document.documentElement.scrollHeight) || document.body.scrollHeight;
    var clientHeight = document.documentElement.clientHeight || window.innerHeight;
    var scrolledToBottom = Math.ceil(scrollTop + clientHeight) >= scrollHeight;

    if (scrolledToBottom && !this.props.isLoading) {
      this.props.loadUsers();
    }
  }
  render() {
    const { users } = this.props;
    return (
      <React.Fragment>
        <Helmet>
          <title>Home</title>
        </Helmet>

        <div className="row center-xs">
          <div className="col-xs-6">
            <div className="box">

              <div className="row start-xs">
                <div className="col-xs-12">
                  <div className="box">
                    <h2>{config('welcomeMessage')}</h2>
                    <Link to="/setting">setting</Link>
                    {users && users.map((value) => (
                      <UsersCard user={value} key={value.email} />
                    ))}
                  </div>
                </div>
              </div>

            </div>
          </div>
        </div>
      </React.Fragment>
    );
  }
}

const mapStateToProps = state => ({
  isError: state.users.data.isError,
  isLoading: state.users.data.isLoading,
  users: state.users.data.results,
});

const mapActionsToProps = {
  loadInitialUsers: loadInitialUsers,
  loadUsers: loadUsers
};

PageHome.propTypes = {};

export default compose(
  connect(mapStateToProps, mapActionsToProps),
  withJob({
    work: (
      { match, users, loadInitialUsers }, // eslint-disable-line
    ) => {
      // naive caching
      if (users.length > 0) {
        return false
      }
      // Execute the redux-thunk powered action that returns a Promise and
      // fetches the users.
      loadInitialUsers()
      // dont need to trigger the work. unless you need re-render the page with new data
      // ex: navigating from (current page) /article/news-1 to /article/news-2
      // we need to update with news-2 data on the same page/component right? then we need
      // to trigger the work and pass news-2 parameter; update redux state tree.
      shouldWorkAgain: (prevProps, nextProps) => {
        // eslint-disable-line
        return false;
      }
    }
  }),
)(PageHome);
