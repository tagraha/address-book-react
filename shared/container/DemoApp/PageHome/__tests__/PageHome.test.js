/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import { shallow } from 'enzyme';
import axios from 'axios';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk'
import { usersMock } from '../../../../redux/initialStates';
import { loadUsers, ACTIONS } from '../../../../redux/modules/users';
import PageHome from '../PageHome';

const middlewares = [thunk.withExtraArgument({ axios })];
const mockStore = configureMockStore(middlewares);
const store = mockStore({users: usersMock});

describe('should render <PageHome /> correctly', () => {
  test('renders', () => {
    const wrapper = shallow(<PageHome store={store} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('it should dispatch action correctly when load the users', () => {
    const expectedActions = [
      { type: ACTIONS.FETCH_USERS },
      { type: ACTIONS.FETCH_USERS_SUCCESS, payload: {} },
    ]
    store.dispatch(loadUsers()).then((args) => {
      console.log('args');
      // ASSERTIONS / EXPECTS
      expect(store.getActions()).toEqual(expectedActions)
    });
  });
});
