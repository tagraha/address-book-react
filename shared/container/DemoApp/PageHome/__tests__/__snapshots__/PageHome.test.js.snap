// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render <PageHome /> correctly renders 1`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Home
    </title>
  </HelmetWrapper>
  <div
    className="homepage-wrapper"
  >
    <div
      className="fixed-wrapper"
    >
      <div
        className="search-wrapper"
      >
        <Connect(FilterInput) />
      </div>
    </div>
    <Connect(WithJob(UsersCards)) />
  </div>
</Fragment>
`;
